### PR TITLE
Tutorial levels

### PR DESCRIPTION
**Pull request by cdsupina**
- Changed input handling to be "action" based use Leafwing Input Manager
- Optional tutorial level
  - phases for tracking tutorial objectives
  - can be toggled on and off in character selection screen with (up and down inputs)
  - only available for single player
  - 3 generic tutorials for movement, attack (and protect), and special ability,
- Overhauled game ui to use hierarchy from single game ui entity

### CODE REVIEW

File: src/ui/game.rs [hunk 1]
```diff
@@ -1,93 +1,554 @@
...
+
+/// Tag for level ui
+#[derive(Component)]
+pub struct PowerGlowUI(Timer);
+
+#[derive(Component)]
+pub struct StatBarLabel;
+
+#[derive(Component)]
+pub struct AbilityChargingUI;
```
Comment (varoonp123) [inline src/ui/game.rs]: Dont make things public unless you expect other plugins/systems to use it. These can be private to the module. Really the only things that should need to be public are systems, and a couple other bevy things.

File: crates/thetawave_interface/src/options/mod.rs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            })*
+        }
+    }
+}
+
+/// Generates the display config file
+pub fn generate_config_files() {
+    confgen!("display.ron");
+    confgen!("input.ron");
+}
```
Comment (varoonp123) [inline crates/thetawave_interface/src/options/mod.rs]: I think we should delete this. CLI parameters are a more standard way to do this. `thetawave --resolution 1920x1080 --fullscreen` and those can be used for launch options in steam. I dont think we want others to change this via ron (99.99% of people dont know what RON is, but almost all programmers can do `thetawave --help` and set CLI params). 

https://github.com/thetawavegame/thetawave/blob/0447952b94077c66cb8bc92fcaa4e210ab307aff/src/options/mod.rs#L26

Atleast outside of the browser, we can save user settings in sqlite. Keeping a couple of preconfigured ones as ron is...fine for now (if they _can_ be toml and are simple enough, I think toml is better and more portable/standard; i dont particularly value serde's data model enough to have a strong preference for ron).

File: crates/thetawave_interface/src/states.rs [hunk 3]
```diff
@@ -20,3 +20,24 @@ pub enum GameStates {
...
+pub struct VictoryCleanup;
+
+#[derive(Component)]
+pub struct PauseCleanup;
+
+#[derive(Component)]
+pub struct InstructionsCleanup;
+
+#[derive(Component)]
+pub struct CharacterSelectionCleanup;
```
Comment (varoonp123) [inline crates/thetawave_interface/src/states.rs]: Maybe 

```rust
pub struct CleanupTag(AppState)
```
or [bit flags](https://stackoverflow.com/questions/43509560/how-to-implement-bitwise-operations-on-a-bitflags-enum) if you want to assign multiple cleanup tags (and back it by a u8)
or move these to a module like `cleanup_tags` with a top level doc like 

```
Minimal indicators for when attributes should be despawned from the game. 
```

File: crates/thetawave_interface/src/options/input.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+use bevy_input::{gamepad::GamepadButtonType, keyboard::KeyCode, mouse::MouseButton};
+use bevy_reflect::Reflect;
+use leafwing_input_manager::{prelude::InputMap, Actionlike};
+use serde::Deserialize;
+
+#[derive(Component)]
+pub struct MenuExplorer;
+
+#[derive(Actionlike, PartialEq, Eq, Clone, Copy, Hash, Debug, Reflect, Deserialize)]
+pub enum MenuAction {
```
Comment (varoonp123) [inline crates/thetawave_interface/src/options/input.rs:11]: +1 and i like how much it cleans up the game code.

File: crates/thetawave_interface/src/objective.rs [hunk 5]
```diff
@@ -5,8 +5,55 @@ use serde::Deserialize;
...
 #[derive(Deserialize, Clone)]
 pub enum DefenseInteraction {
     Heal(usize),
     Damage(usize),
 }
+
+#[derive(Event)]
+pub struct NewObjectiveEvent {
+    pub objective: Option<Objective>,
+}
```
Comment (varoonp123) [inline crates/thetawave_interface/src/objective.rs]: What does `NewObjectiveEvent(None)` mean?

File: crates/thetawave_interface/src/player.rs [hunk 6]
```diff
@@ -1,8 +1,48 @@
...
+    // A method to get a vector of all used inputs
+    pub fn get_used_inputs(&self) -> Vec<PlayerInput> {
+        self.player_data
+            .iter()
+            .filter_map(|player_data| player_data.clone().map(|data| data.input))
+            .collect()
+    }
+}
+
+/// Player input
```
Comment (varoonp123) [inline crates/thetawave_interface/src/player.rs:40]: Brilliant comment

File: src/player/systems/attacks.rs [hunk 7]
```diff
@@ -18,90 +24,82 @@ pub fn player_scale_fire_rate_system(mut player_query: Query<&mut PlayerComponen
...
-        .map(|gamepad| {
-            (
-                gamepad.id,
-                gamepad_input.pressed(GamepadButton {
-                    gamepad,
-                    button_type: GamepadButtonType::RightTrigger,
-                }),
-            )
+    // first make sure that the players can shoot based on the level
+    let can_fire_weapon = run_resource.current_level.clone().is_some_and(|level| {
```
Comment (varoonp123) [inline src/player/systems/attacks.rs]: I think we can avoid cloning this by taking a reference somewhere (probably like `|&level|` and `if let ... = &thing) `

File: crates/thetawave_interface/src/options/mod.rs [hunk 8]
```diff
@@ -0,0 +1,52 @@
+use std::{
```
Comment (varoonp123) [inline crates/thetawave_interface/src/options/mod.rs]: Now that I look at this again, I dont like this in the interface crate. How this stuff is parsed + CLI is detinitely implementation. Let's move this back to the main `src/` crate.

File: crates/thetawave_interface/src/spawnable.rs [hunk 9]
```diff
@@ -152,3 +184,28 @@ pub enum TextEffectType {
...
+
+/// Event for spawning mobs
+#[derive(Event)]
+pub struct SpawnMobEvent {
+    /// Type of mob to spawn
+    pub mob_type: MobType,
+    /// Position to spawn mob
+    pub position: Vec2,
+
+    pub rotation: Quat,
```
Comment (varoonp123) [inline crates/thetawave_interface/src/spawnable.rs]: Why is this a quaternion when the game is 2d?

File: crates/thetawave_interface/src/options/input.rs [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+    MoveUp,
+    MoveDown,
+    MoveLeft,
+    MoveRight,
+    BasicAttack,
+    SpecialAttack,
+}
+
+#[derive(Deserialize)]
+pub struct InputBindings {
```
Comment (varoonp123) [inline crates/thetawave_interface/src/options/input.rs:35]:
```suggestion
struct InputBindings {
```

File: src/player/resources/mod.rs [hunk 11]
```diff
@@ -1,7 +1,5 @@
 //! Resources for managing players
 
 mod character;
-mod players;
 
-pub use self::character::{Character, CharacterType, CharactersResource};
-pub use self::players::{PlayerInput, PlayersResource};
+pub use self::character::CharactersResource;
```
Comment (varoonp123) [inline src/player/resources/mod.rs]: This module now doesn't have anything. Can we delete it/just use this import elsewhere?

File: src/ui/player.rs [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+                        flex_direction: FlexDirection::ColumnReverse,
+                        ..default()
+                    },
+                    ..default()
+                })
+                .insert(ArmorUI(player_index));
+        });
+}
+
+fn build_armor_counter(parent: &mut ChildBuilder) {
```
Comment (varoonp123) [inline src/ui/player.rs:303]:
```suggestion
fn build_armor_counter() -> NodeBundle {
```

File: src/ui/phase.rs [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+                                        }
+                                    }
+                                });
+                        });
+                    }
+                }
+            }
+        }
+    }
+}
```
Comment (varoonp123) [inline src/ui/phase.rs:266]: Gamedev brained ....

File: crates/thetawave_interface/src/states.rs (hunk 3 above)
Comment (cdsupina) [reply crates/thetawave_interface/src/states.rs]: I attempted to generalize the cleanup tag, but was not successful. I think this is still the best way of doing it. I'll move them to a module like you suggested though.

File: assets/sounds/objective_completed.wav
Comment (varoonp123) [inline assets/sounds/objective_completed.wav:1]: The other files are < 1k so i dont really care if they are in git. This sound file should be in S3 though.

File: crates/thetawave_interface/src/player.rs [hunk 14]
```diff
@@ -1,8 +1,48 @@
...
+    pub input: PlayerInput,
+}
+
+impl Default for PlayersResource {
+    fn default() -> Self {
+        PlayersResource {
+            player_data: vec![None, None, None, None],
+        }
+    }
+}
```
Comment (varoonp123) [inline crates/thetawave_interface/src/player.rs:28]: `PlayersResource` doesnt make any sense to me. Why is it Vec<Option<_>>? Why is the default 4 nones?

File: crates/thetawave_interface/src/objective.rs [hunk 15]
```diff
@@ -1,9 +1,65 @@
...
+use crate::spawnable::{MobSegmentType, MobType};
+
 // Event for sending damage dealt from mob reaching bottom of arena
 #[derive(Event)]
-pub struct MobReachedBottomGateEvent(pub DefenseInteraction);
+pub struct MobReachedBottomGateEvent {
+    pub mob_type: Option<MobType>,
+    pub mob_segment_type: Option<MobSegmentType>,
+    pub defense_interaction: DefenseInteraction,
+}
```
Comment (varoonp123) [inline crates/thetawave_interface/src/objective.rs:12]: Hmm. Lets talk about this type. Should this have an enum? Or just an `Entity`? 2 options seems odd.

File: crates/thetawave_interface/src/options/input.rs (hunk 10 above)
Comment (cdsupina) [reply crates/thetawave_interface/src/options/input.rs:35]: This needs to remain public:
https://github.com/thetawavegame/thetawave/pull/108/files#diff-bbff6d4c12e9078362fb1f4999f7ed294baa96c19c6fb33345937573334c1527R47

File: src/ui/phase.rs (hunk 13 above)
Comment (cdsupina) [reply src/ui/phase.rs:266]: OnePieceGear5GIF

File: assets/sounds/objective_completed.wav
Comment (cdsupina) [reply assets/sounds/objective_completed.wav:1]: help

File: crates/thetawave_interface/src/player.rs (hunk 14 above)
Comment (cdsupina) [reply crates/thetawave_interface/src/player.rs:28]: It's for Nones for up to for possible players in the game. When a player slot is filled it becomes Some(PlayerData) for that slot.

**Pull request closed without merging**